server/apps/legal_archive/commands.py

```diff
@@ -0,0 +1,285 @@
+# at https://www.sourcefabric.org/superdesk/license
+
+import logging
+import superdesk
+from copy import deepcopy
+from flask import current_app as app
+from eve.utils import ParsedRequest
+from eve.versioning import versioned_id_field
+from superdesk.celery_app import celery
+from superdesk import get_resource_service, config
+from superdesk.celery_task_utils import get_lock_id
+from .resource import LEGAL_ARCHIVE_NAME, LEGAL_ARCHIVE_VERSIONS_NAME, LEGAL_PUBLISH_QUEUE_NAME
+from superdesk.users.services import get_display_name
+from apps.archive.common import ARCHIVE
+from superdesk.metadata.item import ITEM_STATE, CONTENT_STATE
+from superdesk.lock import lock, unlock
+
+logger = logging.getLogger(__name__)
+
+
+class LegalArchiveImport:
+
+    log_msg_format = "{{'_id': {_id}, 'unique_name': {unique_name}, 'version': {_current_version}, " \
+                     "'expired_on': {expiry}}}."
+
+    def upsert_into_legal_archive(self, doc):
+        """
+        Once publish actions are performed on the article do the below:
+            1.  Check if the legal_archive is up-to-date or not.
+            2.  De-normalize the expired article
+            3.  Fetch version history of article so that version_history_doc[config.VERSION] <= doc[config.VERSION]
+                and De-normalize.
+            4.  Upserting Legal Archive.
+            5.  Inserting Legal Archive Versions.
+        :param dict doc: doc from 'archive' collection.
+        """
+
+        if not doc.get(ITEM_STATE) in {CONTENT_STATE.PUBLISHED, CONTENT_STATE.CORRECTED, CONTENT_STATE.KILLED}:
+            logger.exception('Invalid state: {}. Cannot move the item to legal archive. item: {}'.
+                             format(doc.get(ITEM_STATE), self.log_msg_format.format(**doc)))
+            return
+
+        # required for behave test.
+        legal_archive_doc = deepcopy(doc)
+
+        legal_archive_service = get_resource_service(LEGAL_ARCHIVE_NAME)
+        legal_archive_versions_service = get_resource_service(LEGAL_ARCHIVE_VERSIONS_NAME)
+
+        log_msg = self.log_msg_format.format(**legal_archive_doc)
+        version_id_field = versioned_id_field(app.config['DOMAIN'][ARCHIVE])
+        logger.info('Preparing Article to be inserted into Legal Archive ' + log_msg)
+
+        # Removing Irrelevant properties
+        legal_archive_doc.pop(config.ETAG, None)
+        legal_archive_doc.pop('lock_user', None)
+        legal_archive_doc.pop('lock_session', None)
+        legal_archive_doc.pop('lock_time', None)
+
+        logger.info('Removed irrelevant properties from the article ' + log_msg)
+
+        # Step 1
+        article_in_legal_archive = legal_archive_service.find_one(req=None, _id=legal_archive_doc[config.ID_FIELD])
+
+        # if there a version in that has version greater then current doc then return
+        if article_in_legal_archive and article_in_legal_archive[config.VERSION] >= legal_archive_doc[config.VERSION]:
```

Comment:
Ideally the version in legal archive will never be greater than the doc in archive.